### ISSUE TITLE
Problem with npm build of vega

### ISSUE THREAD

**Issue by ellisonbg** [vega/vega#359]
I am using vega in another npm package and am having some build problems with it. When I try to build everything, I find that vega isn't able to resolve some packages that are declared as dev dependencies but not runtime dependencies. More specifically I had to add the following to my `package.json`:

```
  "dependencies": {
    "browserify-shim": "^3.8.10",
    "browserify-versionify": "^1.0.6",
    "jstransform": "^11.0.3",
    "through": "^2.3.8",
    "vega": "^2.2.4",
    "vega-lite": "^0.7.15",
  },
  "browserify-shim": {
    "d3": "global:d3",
    "d3.layout.cloud": "global:d3.layout.cloud",
    "canvas": "global:canvas",
    "topojson": "global:topojson"
  },
```

To get it to work. Are these meant to be actual dependencies or is something else off?

**Comment by jheer** [comment]
We pull vega from npm all the time without issue, so I suspect this has something to do with your build process. For example, there should be no reason whatsoever that you would include `vega-lite` in the `package.json` for vega proper. (Not to mention that vega-lite doesn't even use Vega 2.0 yet...)

So, which version of Vega are you pulling from npm? And what do you mean by "build everything"? 

Hopefully you can `npm install vega` without issue. I would test that first. The installed files should then include both the compiled js files and source files require'able from node.

**Comment by ellisonbg** [comment]
I am exploring some programmatic APIs for Python to work with vega specs and have them rendered in the jupyter/ipython notebook in different ways. First step is to get vega and vega-lite on a notebook page and integrated into a python package. I am using npm and my own package.json to pull vega and vega-lite into my own JavaScript code that will embed things into the notebook (through browserify).

Here is the package.json I am using:

https://github.com/ellisonbg/vegaexplore/blob/master/package.json

With vega 2.x latest this package.json does work. With vega 1.5 latest it fails. Do you have examples of projects that use vega and vega-lite through npm?

**Comment by jheer** [comment]
Hmm, I see. Honestly, I think the best solution may be to wait a week or two, at which point we should hopefully have moved vega-lite over to vega 2.x. I wouldn't want you to put in a bunch of effort now, then pull the carpet out from under you. Flagging this for @kanitw and @domoritz.

In the meantime, you might explore vega 2.x integration, and follow up with vega-lite once we move it over.

**Comment by domoritz** [comment]
Moving to vega 2 for vega-lite is high on the roadmap (issue at https://github.com/vega/vega-lite/issues/459). 

Also note that we are working on vega/ vega-lite integration into IPython notebook at https://github.com/vega/ipython-vega. I haven't started on the raw vega part yet if you want to contribute.

**Comment by ellisonbg** [comment]
OK I can play some with vega 2 while vega-lite is updated. 

But then the build problems with vega2 are still there (I have to add the extra devDeps oand browserify-shim to *my* package.json). Do you have examples of any package.json files that use vega2?

**Comment by ellisonbg** [comment]
Do you think my build problems are related to my not listing d3 and other
things in my package.json?

On Tue, Sep 15, 2015 at 10:05 PM, Dominik Moritz <notifications@github.com>
wrote:

> Moving to vega 2 for vega-lite is high on the roadmap (issue at
> vega/vega-lite#459 <https://github.com/vega/vega-lite/issues/459>).
>
> Also note that we are working on vega/ vega-lite integration into IPython
> notebook at https://github.com/vega/ipython-vega. I haven't started on
> the raw vega part yet if you want to contribute.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/vega/vega/issues/359#issuecomment-140627352>.
>



-- 
Brian E. Granger
Associate Professor of Physics and Data Science
Cal Poly State University, San Luis Obispo
@ellisonbg on Twitter and GitHub
bgranger@calpoly.edu and ellisonbg@gmail.com

**Comment by arvind** [comment]
Hi Brian,

Could you describe the build problems more specifically (e.g., error
messages when you execute npm install?). That would help track this down
further.

On Wednesday, September 16, 2015, Brian E. Granger <notifications@github.com>
wrote:

> Do you think my build problems are related to my not listing d3 and other
> things in my package.json?
>
> On Tue, Sep 15, 2015 at 10:05 PM, Dominik Moritz <notifications@github.com
> <javascript:_e(%7B%7D,'cvml','notifications@github.com');>>
> wrote:
>
> > Moving to vega 2 for vega-lite is high on the roadmap (issue at
> > vega/vega-lite#459 <https://github.com/vega/vega-lite/issues/459>).
> >
> > Also note that we are working on vega/ vega-lite integration into IPython
> > notebook at https://github.com/vega/ipython-vega. I haven't started on
> > the raw vega part yet if you want to contribute.
> >
> > —
> > Reply to this email directly or view it on GitHub
> > <https://github.com/vega/vega/issues/359#issuecomment-140627352>.
> >
>
>
>
> --
> Brian E. Granger
> Associate Professor of Physics and Data Science
> Cal Poly State University, San Luis Obispo
> @ellisonbg on Twitter and GitHub
> bgranger@calpoly.edu
> <javascript:_e(%7B%7D,'cvml','bgranger@calpoly.edu');> and
> ellisonbg@gmail.com <javascript:_e(%7B%7D,'cvml','ellisonbg@gmail.com');>
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/vega/vega/issues/359#issuecomment-140786705>.
>


-- 
-Arvind
-----
Arvind Satyanarayan <http://arvindsatya.com/>

**Comment by meetamit** [comment]
I think I'm running into the same issue. I'm using browserify to build a js file for a project, with Vega (along with React etc) being one of the dependencies. I.e, somewhere in my code I have `require('vega')`.

When browserify builds my project it throws an error. The error is

    Error: Cannot find module 'jstransform'

If comes from `/node_modules/vega/scripts/strip-schema.js`. I.e. when browserify builds my package, `strip-schema.js` gets called, which is surprising to me (but maybe that makes sense in browserify land). Since `jstransform` is dependency of `strip-schema` but is listed as dev dependency of vega, it doesn't get installed automatically upon `npm install vega`.

So I tried

    cd node_modules/vega
    npm install --dev

Which installed `jstransform` and the rest of the dev dependencies. Although that took away the missing module error, a new error appeared:

    Error: Unable to find a browserify-shim config section in the package.json for [redacted]/node_modules/browserify/node_modules/browser-resolve/empty.js while parsing file: [redacted]/node_modules/browserify/node_modules/browser-resolve/empty.js
      at [redacted]/node_modules/vega/node_modules/browserify-shim/lib/resolve-shims.js:163:38
      at [redacted]/node_modules/vega/node_modules/browserify-shim/node_modules/mothership/index.js:27:31
      at testDir ([redacted]/node_modules/vega/node_modules/browserify-shim/node_modules/mothership/node_modules/find-parent-dir/index.js:20:36)
      at [redacted]/node_modules/vega/node_modules/browserify-shim/node_modules/mothership/node_modules/find-parent-dir/index.js:26:7
      at FSReqWrap.cb [as oncomplete] (fs.js:226:19)

Bottom line, it seems like vega cannot be browserify'ed.

**Comment by arvind** [comment]
Ah interesting, I see what's happening. In our other projects that `require('vega')` and use browserify (e.g., [vega-embed](https://github.com/vega/vega-embed/blob/master/package.json)), Vega is listed as an external dependency (i.e., `browserify-shim.vega = 'global:vg') and as a result doesn't get packaged up by browserify again. 

Now the question is, how exactly do we resolve this situation? Is the right thing to do to move all the packages required for building vega (browserify, browserify-shim, browserify-versionify, jstransform, etc.) to full dependencies? Or, is it reasonable to expect users to install Vega devDependencies if you wish to build Vega as part of your own package? I'm not sure what the npm/browserify precedent here is.

**Comment by meetamit** [comment]
Agreed. I don't have an obvious solution either. I was hoping maybe there would be some "magic" browserify options I could add to my own `package.json` that would make browserify incorporate vega.js without needing to run any dev-related tasks, but couldn't come up with anything.

Note that one could take vega out of one's own browserify'ed build (and instead include it as a `<script>` tag in the DOM, but that would entail also taking d3 out of the browserify build.

**Comment by arvind** [comment]
So, investigating this further, here's what I have so far. In a stub project that defines vega as a dependency or devDependency, to have Vega be correctly included as part of the browserify package, I need to:

1. `cd node_modules/vega; npm install;`
2. Add `"browserify-shim": {}` to my stub project's `package.json`.

**Comment by jheer** [comment]
Can it be made to work by instead adding the required devDependencies to the _top_ level package.json? It seems reasonable to include those as devDependencies for the top level project since the goal is to create a unified bundle... and it would be include to note this as an option if feasible.

Either way, we should add some documentation to the wiki about this! Adding the documentation tag...

**Comment by meetamit** [comment]
Now documented https://github.com/vega/vega/wiki/Vega-and-Browserify

**Comment by arvind** [comment]
:+1: Thanks a lot @meetamit!

**Comment by jheer** [comment]
Awesome. Thanks @meetamit!

**Comment by meetamit** [comment]
Thank *you* guys

**Comment by meetamit** [comment]
Sorry, but there's still an issue here, although I'm not sure to what extent this is related to browserify.

So I'm able to build vega into a browserify bundle without error. Then, when trying to parse and render a simple spec a runtime error occurs, during the first call to `.update()` on the View. The error itself is

    Uncaught TypeError: Cannot read property '_id' of undefined
      idMap @ Tuple.js:13
      join @ Builder.js:257
      proto.evaluate @ Builder.js:214
      proto.evaluate @ GroupBuilder.js:63
      prototype.evaluate @ Graph.js:275
      prototype.propagate @ Graph.js:127
      prototype.fire @ Model.js:133
      prototype.update @ View.js:340
      (anonymous function) @ (index):18
      (anonymous function) @ spec.js:31

Digging in, I see that in `Builder.js` line 257, which is

    output = join.call(this, input, data);

`data` should be `[{sentinel: 1}]` but is in fact `[undefined]`.

Digging further, I see that on line 10 in that same file, there is

    Sentinel = df.Sentinel

where `df = require('vega-dataflow'),`

However, when bundled with browserify, `df.Sentinel` is undefined. `df` itself is NOT empty. It contains

    {
      ChangeSet: Object
      Collector: Collector(graph)
      DataSource: DataSource(graph, name, facet)
      Dependencies: Object
      Graph: Graph()
      Node: Node(graph)
      Signal: Signal(graph, name, initialValue)
      Tuple: Object
      debug: debug(input, args)
    }

but not `Sentinel`. So apparently `vega-dataflow` was `require`'d but not completely.

I looked further into `vega-dataflow`, and indeed `Sentinel` is not in there, but perhaps expectedly. I could not figure out how in the normal build of vega `{sentinel: 1`} makes its way into the source code. Any hints?

Thanks!

**Comment by arvind** [comment]
This looks like a case of mis-matched versions of vega and vega-dataflow modules. I'd recommend grabbing the latest versions of both with npm install, and you should see that particular error resolve. Sentinel is no longer part of vega-dataflow, and `Builder.js` no longer requires it.

**Comment by meetamit** [comment]
Oops, I was convinced I was running the most recent, but in fact was 2.1. Works perfectly now. Thanks and sorry for the false alarm.

**Comment by jimallman** [comment]
FYI @meetamit (and others finding this issue via Google): I've added a hint to the [Vega-and-Browserify wiki page](https://github.com/vega/vega/wiki/Vega-and-Browserify) to install `watchify` locally, to avoid the missing shim-config errors. Change is [here](https://github.com/vega/vega/wiki/Vega-and-Browserify/_compare/c79ab87cb44868caed32347eb5ef46369e410ff8...e89be988311827e53d0648a1f6e090ed067cab0e) if you're interested.

**Comment by rgbkrk** [comment]
In a similar vein, if `vega` is installed for an Electron app it tries to build canvas, even though it doesn't need it since it will have canvas available. Could a node specific install require a manually installed `canvas` instead?
